Cancel running Flink job Failed（Is cancel running flink job  must use with the savepoint）

Is cancel running flink job  must use with the savepoint？Or cancel running flink job only used in update?
When use  `kubectl annotate flinkclusters <CLUSTER-NAME> flinkclusters.flinkoperator.k8s.io/user-control=job-cancel` to cancel a jobcluster without setting the savepointsDir，the flink-operator-controller-manager failed with the log

`{"level":"info","ts":1642668111.7550068,"logger":"controllers.FlinkCluster","msg":"Trigger savepoint for update","cluster":"default/flinkjobcluster-sample","jobID":"8e964772d5b0ec2860567f82703523cf"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1436a8b]

goroutine 306 [running]:
github.com/spotify/flink-on-k8s-operator/controllers.(*ClusterReconciler).triggerSavepoint(0xc0015b0300, {0xc00159d320, 0x20}, {0x17acabf, 0x6}, 0x0)
        /workspace/controllers/flinkcluster_reconciler.go:846 +0x1ab
github.com/spotify/flink-on-k8s-operator/controllers.(*ClusterReconciler).trySuspendJob(0xc0015b0300)
        /workspace/controllers/flinkcluster_reconciler.go:660 +0x13b
github.com/spotify/flink-on-k8s-operator/controllers.(*ClusterReconciler).reconcileJob(0xc0015b0300)
        /workspace/controllers/flinkcluster_reconciler.go:594 +0xdbc
github.com/spotify/flink-on-k8s-operator/controllers.(*ClusterReconciler).reconcile(0xc0015b0300)
        /workspace/controllers/flinkcluster_reconciler.go:111 +0x189
github.com/spotify/flink-on-k8s-operator/controllers.(*FlinkClusterHandler).reconcile(0xc001051c18, {0x17b2411, 0x4}, {{{0xc0014fb3c9, 0x25d9d20}, {0xc00149fdb8, 0x203000}}})
        /workspace/controllers/flinkcluster_controller.go:223 +0xca5
github.com/spotify/flink-on-k8s-operator/controllers.(*FlinkClusterReconciler).Reconcile(0xc0008a2340, {0x19bf0b8, 0xc0015ba4e0}, {{{0xc0014fb3c9, 0x7}, {0xc00149fdb8, 0x16}}})
        /workspace/controllers/flinkcluster_controller.go:84 +0x30d
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000382a00, {0x19bf010, 0xc00093c000}, {0x162dec0, 0xc0015b69c0})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.6/pkg/internal/controller/controller.go:298 +0x303
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000382a00, {0x19bf010, 0xc00093c000})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.6/pkg/internal/controller/controller.go:253 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.6/pkg/internal/controller/controller.go:214 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2`

--- a/controllers/flinkcluster/flinkcluster_reconciler.go
+++ b/controllers/flinkcluster/flinkcluster_reconciler.go
@@ -593,7 +593,10 @@ func (reconciler *ClusterReconciler) reconcileJob() (ctrl.Result, error) {
 	// Job cancel requested. Stop Flink job.
 	if desiredJob == nil && job.IsActive() {
 		log.Info("Stopping job", "jobID", jobID)
-		newSavepointStatus, err = reconciler.trySuspendJob()
+		if err := reconciler.cancelRunningJobs(true /* takeSavepoint */); err != nil {
+			return requeueResult, err
+		}
+
 		var userControl = getNewControlRequest(observed.cluster)
 		if userControl == v1beta1.ControlNameJobCancel {
 			newControlStatus = getControlStatus(userControl, v1beta1.ControlStateInProgress)
@@ -653,6 +656,11 @@ func (reconciler *ClusterReconciler) getFlinkJobID() string {
 func (reconciler *ClusterReconciler) trySuspendJob() (*v1beta1.SavepointStatus, error) {
 	var log = reconciler.log
 	var recorded = reconciler.observed.cluster.Status
+
+	if !canTakeSavepoint(*reconciler.observed.cluster) {
+		return nil, nil
+	}
+
 	var jobID = reconciler.getFlinkJobID()
 
 	log.Info("Checking the conditions for progressing")
@@ -705,7 +713,6 @@ func (reconciler *ClusterReconciler) cancelUnexpectedJobs(
 // Cancel running jobs.
 func (reconciler *ClusterReconciler) cancelRunningJobs(
 	takeSavepoint bool) error {
-	var log = reconciler.log
 	var runningJobs = reconciler.observed.flinkJob.unexpected
 	var flinkJob = reconciler.observed.flinkJob.status
 	if flinkJob != nil && flinkJob.Id != "" &&
@@ -713,8 +720,7 @@ func (reconciler *ClusterReconciler) cancelRunningJobs(
 		runningJobs = append(runningJobs, flinkJob.Id)
 	}
 	if len(runningJobs) == 0 {
-		log.Info("No running Flink jobs to stop.")
-		return nil
+		return fmt.Errorf("no running Flink jobs to stop")
 	}
 	return reconciler.cancelJobs(takeSavepoint, runningJobs)
 }

--- a/controllers/flinkcluster/flinkcluster_updater.go
+++ b/controllers/flinkcluster/flinkcluster_updater.go
@@ -903,17 +903,27 @@ func deriveControlStatus(
 		c = recordedControl.DeepCopy()
 		switch recordedControl.Name {
 		case v1beta1.ControlNameJobCancel:
-			if newSavepoint.State == v1beta1.SavepointStateSucceeded && newJob.State == v1beta1.JobStateCancelled {
-				c.State = v1beta1.ControlStateSucceeded
-			} else if newJob.IsStopped() {
-				c.Message = "Aborted job cancellation: savepoint is not completed, but job is stopped already."
-				c.State = v1beta1.ControlStateFailed
-			} else if newSavepoint.IsFailed() && newSavepoint.TriggerReason == v1beta1.SavepointReasonJobCancel {
-				c.Message = "Aborted job cancellation: failed to take savepoint."
+			switch {
+			case newJob.State == v1beta1.JobStateCancelled:
+				if newSavepoint != nil {
+					if newSavepoint.State == v1beta1.SavepointStateSucceeded {
+						c.State = v1beta1.ControlStateSucceeded
+					} else if newSavepoint.IsFailed() && newSavepoint.TriggerReason == v1beta1.SavepointReasonJobCancel {
+						c.Message = "Aborted job cancellation: failed to take savepoint."
+						c.State = v1beta1.ControlStateFailed
+					}
+				} else {
+					c.State = v1beta1.ControlStateSucceeded
+				}
+			case newJob.IsStopped():
+				c.Message = "Aborted job cancellation: job is stopped already."
 				c.State = v1beta1.ControlStateFailed
 			}
 		case v1beta1.ControlNameSavepoint:
-			if newSavepoint.State == v1beta1.SavepointStateSucceeded {
+			if newSavepoint == nil {
+				c.Message = "Aborted: savepoint not defined"
+				c.State = v1beta1.ControlStateFailed
+			} else if newSavepoint.State == v1beta1.SavepointStateSucceeded {
 				c.State = v1beta1.ControlStateSucceeded
 			} else if newSavepoint.IsFailed() && newSavepoint.TriggerReason == v1beta1.SavepointReasonUserRequested {
 				c.State = v1beta1.ControlStateFailed
@@ -925,6 +935,7 @@ func deriveControlStatus(
 		}
 		return c
 	}
+
 	// Maintain control status if there is no change.
 	if recordedControl != nil && c == nil {
 		c = recordedControl.DeepCopy()

